File: day-02/solver.go
package main

import (
    "bufio"
    "os"
    "strings"
    "strconv"
    "fmt"
)


func readData() []string {
    f, _ := os.Open("input.txt")
    defer f.Close()

    data := make([]string, 0)

    scanner := bufio.NewScanner(f)

    for scanner.Scan() {
        data = append(data, scanner.Text())
    }

    return data
}

func partOne(horiz int, depth int, line string) (int, int) {
    pieces := strings.Fields(line)
    verb := pieces[0]
    amount, _ := strconv.Atoi(pieces[1])
    switch verb {
    case "forward":
        horiz += amount
    case "down":
        depth += amount
    case "up":
        depth -= amount
    }
    return horiz, depth
}

func partTwo(horiz int, depth int, aim int, line string) (int, int, int) {
    pieces := strings.Fields(line)
    verb := pieces[0]
    amount, _ := strconv.Atoi(pieces[1])
    switch verb {
    case "forward":
        horiz += amount
        depth += aim * amount
    case "down":
        aim += amount
    case "up":
        aim -= amount
    }
    return horiz, depth, aim
}

func main() {
    data := readData()

    // Part 1
    horiz := 0
    depth := 0
    for _, line := range data {
        horiz, depth = partOne(horiz, depth, line)
    }
    fmt.Println(horiz * depth)

    // Part 2
    horiz = 0
    depth = 0
    aim := 0
    for _, line := range data {
        horiz, depth, aim = partTwo(horiz, depth, aim, line)
    }
    fmt.Println(horiz * depth)

}

File: day-02/solver.py
data = [x for x in open('input.txt').read().split('\n')]

# Part 1

horiz = depth = 0

verb_map = {
    'forward' : lambda x: (horiz + x, depth),
    'down': lambda x: (horiz, depth + x),
    'up': lambda x: (horiz, depth - x),
}

for datum in data:
    verb, amount = datum.split(' ')
    horiz, depth = verb_map[verb](int(amount))

print(horiz * depth)

# Part 2

aim = horiz = depth = 0

verb_map = {
    'forward' : lambda x: (horiz + x, depth + (aim * x), aim),
    'down': lambda x: (horiz, depth, aim + x),
    'up': lambda x: (horiz, depth, aim - x),
}

for datum in data:
    verb, amount = datum.split(' ')
    horiz, depth, aim = verb_map[verb](int(amount))

print(horiz * depth)
